### PR TITLE
Update docs and hooks worker client for go hooks

### PR DESCRIPTION
**Pull request by ddelnano**
This PR depends on [this](https://github.com/snikch/goodman/pull/5).  Once that is merged, the go hooks should be considered compliant with the hooks for other languages: php, ruby, python, perl, etc.

I tested the changes with this [repo](https://github.com/ddelnano/dredd-hooks-go-example).  I could not install dredd from a github repo since npm does not run prepublish and the necessary commands to do so I made these commits on my fork and made sure the examples worked.

screen shot 2016-05-31 at 10 58 52 pm

### CODE REVIEW

**Comment by ddelnano** [comment]
@honzajavorek The PR this depended on is finally [merged](https://github.com/snikch/goodman/pull/5)

**Comment by honzajavorek** [comment]
@ddelnano Great! I asked our Go lover @pksunkara to look at it once he has a free moment, so expect him to land & comment here soon.

**Comment by ddelnano** [comment]
@honzajavorek I am having problems with the Go hooks dealing with requests that have an empty body.  Like the ones in the hooks template.  If I add a request body all the aruba/cucumber tests pass.  If I don't I get this dredd error.
screen shot 2016-06-03 at 9 00 04 am

Seems to me to be kind of like the error in #499 where it shouldn't allow whatever value that is to be passed to Buffer.byteLength, but I think there is some other issue where maybe the go hooks send back bad data or something that causes dredd to have these errors.

**Comment by ddelnano** [comment]
@honzajavorek so the CI build for the go hooks pass if I add this to every .feature file

https://github.com/snikch/goodman/blob/master/features/execution_order.feature#L19-L20

**Comment by ddelnano** [comment]
@honzajavorek deleted my previous comments, cause I thought this was on an issue and not this PR.  Sounds good to me 👍

File: docs/hooks-go.md [hunk 1]
```diff
@@ -10,21 +10,49 @@ Go hooks are using [Dredd's hooks handler socket interface](hooks-new-language.m
...
 ```
 $ go get github.com/snikch/goodman
+$ cd $GOPATH/src/github.com/snikch/goodman
+$ go build -o $GOPATH/bin/goodman github.com/snikch/goodman/bin
 ```
 
 ## Usage
 
-Using Dredd with Go is slightly different to other languages, as a binary needs to be compiled for execution. Instead of providing a variety of hookfiles to run, your own binary is built using the Go hooks library `goodman` to run your specific hooks. The path to this binary is passed to dredd via the `--language` flag. You may also need to supply a `--hookfiles` flag to get the tests to run, but the value of this is of no matter (the dredd runner requires it).
+Using Dredd with Go is slightly different to other languages, as a binary needs to be compiled for execution. The --hookfiles flags will be compiled hook binaries.  See below for an example hooks.go file to get an idea of what this would look like.
```
Comment (pksunkara) [inline docs/hooks-go.md]: > hookfiles flags should point to compiled hook binaries.

> get an idea of what the source file behind the go binary would look like.

File: docs/hooks-go.md [hunk 2]
```diff
@@ -10,21 +10,49 @@ Go hooks are using [Dredd's hooks handler socket interface](hooks-new-language.m
...
 
 ```
-$ dredd apiary.apib http://localhost:3000 --language ./my-dredd-hook-server --hookfiles *.go
+$ dredd apiary.apib http://localhost:3000 --server ./go-lang-web-server-to-test --language go --hookfiles ./hook-file-binary
```
Comment (pksunkara) [inline docs/hooks-go.md]: @honzajavorek What is this server option used for? I thought localhost:3000 is the server.

Comment (pksunkara) [inline docs/hooks-go.md]: @ddelnano If you are talking about hook server, maybe rename it to `./golang-hook-server`

Comment (ddelnano) [inline docs/hooks-go.md]: @pksunkara the --server option is the command to get the server that will be under test started.  So the command assumes that the `./go-lang-web-server` is a go web service that will start a webserver that listens on localhost port 3000.

Comment (ddelnano) [inline docs/hooks-go.md]: @pksunkara That is not referring to the hook server.

Comment (pksunkara) [inline docs/hooks-go.md]: Ah, thanks. I didn't know that.

Comment (ddelnano) [inline docs/hooks-go.md]: @pksunkara the hooks server command gets set [here](https://github.com/apiaryio/dredd/pull/505/files#diff-fab7c8fd00434dbb9259f70a29405350R138).

File: docs/hooks-go.md [hunk 3]
```diff
@@ -130,21 +156,22 @@ package main
...
-      t.Fail = true
+    h := hooks.NewHooks()
+    server := hooks.NewServer(h)
+    h.Before("Machines > Machines collection > Get Machines", func(t *trans.Transaction) {
+        t.Fail = true
     })
-    runner.Before("Machines > Machines collection > Post Machines", func(t *goodman.Transaction) {
-      t.Fail = "POST is broken"
+    h.Before("Machines > Machines collection > Post  Machines", func(t *trans.Transaction) {
+        t.Fail = "Post is broken"
```
Comment (pksunkara) [inline docs/hooks-go.md]: `POST`

**Comment by pksunkara** [comment]
Just 2 minor comments regarding docs. Will merge it after they are fixed.

File: docs/hooks-go.md (hunk 3 above)
Comment (ddelnano) [inline docs/hooks-go.md]: @pksunkara updated.

File: docs/hooks-go.md (hunk 1 above)
Comment (ddelnano) [inline docs/hooks-go.md]: @pksunkara can you be more specific?  Is this just not clear?

Comment (pksunkara) [inline docs/hooks-go.md]: I just made the sentences a bit more clear. My comment has the updated sentences which you need to copy.

Comment (ddelnano) [inline docs/hooks-go.md]: @pksunkara awesome thanks for clarifying.  I updated it.

**Comment by ddelnano** [comment]
@pksunkara @honzajavorek I updated the PR addressing the feedback **before** this get merged, however, I am thinking there might be something that might need addressed first.  Dredd crashes with the Go hooks if the API Blueprint transaction has an empty request body.  See the difference between the Go [tests](https://github.com/snikch/goodman/blob/master/features/execution_order.feature#L19-L20) and the template [tests](https://github.com/apiaryio/dredd-hooks-template/blob/master/features/execution_order.feature#L16-L20)

**Comment by ddelnano** [comment]
Here is an example of the dredd crash, I am going to try and prepare a fix.

screen shot 2016-06-03 at 1 27 54 pm

**Comment by ddelnano** [comment]
@honzajavorek not sure exactly what the deal is with the findings in my previous comments.  I wrote out a full example using a sample go app, goodman, and dredd compiled from this branch and blueprint files that have empty request bodies run fine.  You can see the example [here](https://github.com/snikch/goodman/pull/7).  Either of you could run the example if you wanted to following the instructions in the Wiki to verify my claims that it works as expected, however, you would need to update the path in the [Makefile](https://github.com/snikch/goodman/pull/7/files#diff-e8849b97cf69febe9876e8ab47d44460R10) since I needed to use dredd compiled from this branch (since it uses these changes).

Let me know what you think.

**Comment by pksunkara** [comment]
Does the above crash happen with other language hook files or only golang? Depending on that, we need to fix it in a separate PR. I am merging this for now.

**Comment by ddelnano** [comment]
@pksunkara when I developed the php hooks I never ran into this.  I also tried to reproduce it with a real example instead of the cucumber tests (where I was seeing it) and I was unable to do so.  I opened a PR for another crash that I found that indicates a hook server failure ( #511 ).  This would just provide better information to the hook server dev instead of seeing the error `cannot read property 'length' of undefined`.

**Comment by ddelnano** [comment]
@pksunkara @honzajavorek when do the pre releases go out?  I would like to merge my example in the goodman repo but since npm's support for installing repos via commit hashes does not run any building (so the coffeescript would not be compiled) I can't merge until this change is in a new release.

**Comment by honzajavorek** [comment]
@ddelnano I should release another one within a day.

**Comment by ddelnano** [comment]
awesome thanks!

**Pull request merged**